File: 0542-01-matrix/0542-01-matrix.java
class Solution {
    public int[][] updateMatrix(int[][] mat) {
        int row = mat.length;
        int col = mat[0].length;
        int[][] dist = new int[row][col];
        Queue<int[]> queue = new ArrayDeque<>();
        
        // Initialize dist array and queue
        for (int i = 0; i < row; i++) {
            for (int j = 0; j < col; j++) {
                if (mat[i][j] == 0) {
                    dist[i][j] = 0;
                    queue.offer(new int[]{i, j});
                } else {
                    dist[i][j] = Integer.MAX_VALUE;
                }
            }
        }
        
        // Directions: up, down, left, right
        int[][] directions = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
        
        // BFS
        while (!queue.isEmpty()) {
            int[] curr = queue.poll();
            int r = curr[0];
            int c = curr[1];
            
            for (int[] dir : directions) {
                int newRow = r + dir[0];
                int newCol = c + dir[1];
                
                if (newRow >= 0 && newRow < row && 
                    newCol >= 0 && newCol < col) {
                    if (dist[newRow][newCol] > dist[r][c] + 1) {
                        dist[newRow][newCol] = dist[r][c] + 1;
                        queue.offer(new int[]{newRow, newCol});
                    }
                }
            }
        }
        
        return dist;
    }
}

File: 0542-01-matrix/0542-01-matrix.py
class Solution:
    def updateMatrix(self, mat: List[List[int]]) -> List[List[int]]:
        row = len(mat)
        col = len(mat[0])
        dist = [[float('inf')] * col for i in range(row)]
        queue = deque([])
        for i in range(row):
            for j in range(col):
                if mat[i][j] == 0:
                    dist[i][j] = 0
                    queue.append((i,j))
        directions = [(-1,0),(1,0),(0,-1),(0,1)]
        while queue:
            r,c = queue.popleft()
            for x,y in directions:
                new_row = x + r
                new_col = y + c
                if 0 <= new_row < row and 0 <=new_col < col and dist[new_row][new_col] == float('inf'):#if dist[new_row][new_col] == float'inf' -> it means this cell还没有被访问过！
                        dist[new_row][new_col] = dist[r][c] + 1
                        queue.append((new_row, new_col))
        return dist
            
        